### PR TITLE
in context localization for markdown files

### PR DESCRIPTION
**Pull request by pelikhan**
This change adds support for rendering localized markdown files in localhost (downloaded from our cloud) and support for incontext translations using crowdin.
Required:
- [x] move all the crowdin stuff into its own file
- [ ] backend support for "?translate=1" route
- [ ] backend support for @crowdinproject@ macro

Journey: user clicks on "translate this page" link at the bottom of docs page
In Context translations are loaded
image
image

### CODE REVIEW

File: cli/server.ts [hunk 1]
```diff
@@ -1102,26 +1111,33 @@ export function serveAsync(options: ServeOptions) {
...
-            const mdopts = <pxt.docs.RenderOptions>{
-                template: expandDocFileTemplate("docs.html"),
-                markdown: md,
-                theme: pxt.appTarget.appTheme,
-                filepath: pathname,
-                TOC: resolveTOC(pathname)
-            };
-            let html = pxt.docs.renderMarkdown(mdopts)
-            sendHtml(html, U.startsWith(md, "# Not found") ? 404 : 200)
+            const lang = opts["translate"] ? "pxt" : opts["lang"] as string;
```
Comment (pelikhan) [inline cli/server.ts]: @darzu backend support needed for this route

File: cli/server.ts [hunk 2]
```diff
@@ -1102,26 +1111,33 @@ export function serveAsync(options: ServeOptions) {
...
+                        markdown: md,
+                        theme: pxt.appTarget.appTheme,
+                        filepath: pathname,
+                        TOC: resolveTOC(pathname),
+                        pubinfo: {
+                            locale: lang
+                        }
+                    };
+                    if (opts["translate"])
+                        mdopts.pubinfo["incontexttranslations"] = "1";
```
Comment (pelikhan) [inline cli/server.ts]: @darzu backend support to enable this macro

File: pxtlib/emitter/cloud.ts [hunk 3]
```diff
@@ -125,7 +125,7 @@ namespace pxt.Cloud {
             url = `md/${pxt.appTarget.id}/${docid.replace(/^\//, "")}?targetVersion=${encodeURIComponent(targetVersion)}`;
         }
         if (!packaged && locale != "en") {
-            url += `&lang=${encodeURIComponent(Util.userLanguage())}`
+            url += `&lang=${encodeURIComponent(locale)}`
```
Comment (pelikhan) [inline pxtlib/emitter/cloud.ts]: this was a silent bug

File: docfiles/footer.html [hunk 4]
```diff
@@ -22,6 +22,9 @@
             <a class="item" href="https://makecode.com/privacy" target="_blank" rel="noopener">Privacy &amp; Cookies</a>
             <a class="item" href="https://makecode.com/termsofuse" target="_blank" rel="noopener"> Terms Of Use</a>
             <a class="item" href="https://makecode.com/trademarks" target="_blank" rel="noopener">Trademarks</a>
+            <!-- @ifdef crowdinproject -->
```
Comment (pelikhan) [inline docfiles/footer.html]: @darzu need support for this macro value

File: docfiles/docs.html [hunk 5]
```diff
@@ -6,6 +6,13 @@
 	<title>@name@</title>
 	<meta name="Description" content="@description@" />
 
+	<!-- @ifdef incontexttranslations -->
+	<script type="text/javascript">
+		var _jipt = [];
+		_jipt.push(['project', '@crowdinproject@']);
```
Comment (pelikhan) [inline docfiles/docs.html]: need support for @crowdinproject@

File: cli/server.ts (hunk 1 above)
Comment (darzu) [reply cli/server.ts]: So if I'm understanding this right, one of the backend changes needed is simply:
If request comes in with translate=1 q.s., then set "incontexttranslations" in pubinfo

It looks like the backend uses renderMarkdown from backend utils which has logic to turn on/off sections via @ifdef

Comment (pelikhan) [reply cli/server.ts]: Yes correct. The backends need to set incontexttranslations is translate == 1 and use the "pxt" locale

**Pull request merged**